File: typescript/examples/getting-started/03b-tools-stream.ts
// Tools: streaming with tool_call chunks
// Defines a tool and streams the response to see tool_call_start and tool_call_delta chunks.
import { z } from "zod";
import { Opper } from "../../src/index.js";

const opper = new Opper();

console.log("Streaming with tools:");

for await (const chunk of opper.stream("sdk-test-tool-use-stream", {
  input_schema: z.object({
    question: z.string().describe("The user's question"),
  }),
  output_schema: z.object({
    answer: z.string().optional().describe("The assistant's text response"),
    tool_calls: z.array(z.object({
      name: z.string(),
      arguments: z.record(z.string(), z.unknown()),
    })).optional().describe("Tool calls requested by the model"),
  }),
  input: { question: "What is the current weather in Stockholm?" },
  model: "anthropic/claude-sonnet-4.6",
  tools: [
    {
      name: "get_weather",
      description: "Get the current weather for a city",
      parameters: z.object({
        city: z.string().describe("City name"),
        unit: z.enum(["celsius", "fahrenheit"]).optional().describe("Temperature unit"),
      }),
    },
  ],
})) {
  switch (chunk.type) {
    case "tool_call_start":
      console.log(`\n[tool_call_start] name=${chunk.tool_call_name} id=${chunk.tool_call_id}`);
      break;
    case "tool_call_delta":
      process.stdout.write(chunk.tool_call_args);
      break;
    case "content":
      process.stdout.write(chunk.delta);
      break;
    case "done":
      console.log("\n[done]", chunk.usage);
      break;
    case "complete":
      console.log("[complete] data:", JSON.stringify(chunk.data, null, 2));
      break;
    case "error":
      console.error("[error]", chunk.error);
      break;
  }
}

File: python/examples/getting-started/03b_tools_stream.py
# Tools: streaming with tool_call chunks
# Shows raw dict schemas — no Pydantic needed. See 03a for Pydantic approach.
import sys

from opperai import Opper

opper = Opper()

print("Streaming with tools:")

for chunk in opper.stream(
    "sdk-test-tool-use-stream",
    input={"question": "What is the current weather in Stockholm?"},
    input_schema={"type": "object", "properties": {"question": {"type": "string"}}},
    output_schema={
        "type": "object",
        "properties": {
            "answer": {"type": "string"},
            "tool_calls": {"type": "array", "items": {"type": "object"}},
        },
    },
    model="anthropic/claude-sonnet-4.6",
    tools=[
        {
            "name": "get_weather",
            "description": "Get the current weather for a city",
            "parameters": {
                "type": "object",
                "properties": {
                    "city": {"type": "string", "description": "City name"},
                    "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
                },
                "required": ["city"],
            },
        },
    ],
):
    match chunk.type:
        case "tool_call_start":
            print(f"\n[tool_call_start] name={chunk.tool_call_name} id={chunk.tool_call_id}")
        case "tool_call_delta":
            sys.stdout.write(chunk.tool_call_args)
            sys.stdout.flush()
        case "content":
            sys.stdout.write(chunk.delta)
            sys.stdout.flush()
        case "done":
            print(f"\n[done] {chunk.usage}")
        case "complete":
            print(f"[complete] data: {chunk.data}")
        case "error":
            print(f"[error] {chunk.error}", file=sys.stderr)
